File: slp/backend-dotnet/Features/Metrics/RedisMetricsCollector.cs
using StackExchange.Redis;

namespace backend_dotnet.Features.Metrics;

public class RedisMetricsCollector : IMetricsCollector
{
    private readonly IConnectionMultiplexer _redis;
    private readonly ILogger<RedisMetricsCollector> _logger;
    private readonly CircuitBreaker _circuitBreaker;

    public RedisMetricsCollector(
        IConnectionMultiplexer redis,
        ILogger<RedisMetricsCollector> logger)
    {
        _redis = redis;
        _logger = logger;
        _circuitBreaker = new CircuitBreaker(5, TimeSpan.FromMinutes(1), logger);
    }

    public async Task RecordRequestAsync(
        string path, string method, int statusCode, double latencyMs)
    {
        // Kiểm tra circuit breaker trước khi thực hiện
        if (!_circuitBreaker.AllowRequest())
        {
            _logger.LogDebug("Metrics circuit is open, skipping record.");
            return;
        }

        try
        {
            var db = _redis.GetDatabase();
            var bucket = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm");

            var requestKey = $"metric:requests:{bucket}";
            var latencyKey = $"metric:latency:{bucket}";

            var batch = db.CreateBatch();
            var t1 = batch.StringIncrementAsync(requestKey);
            var t2 = batch.ListRightPushAsync(latencyKey, latencyMs.ToString("F2"));

            Task? t3 = null;
            if (statusCode >= 400)
                t3 = batch.StringIncrementAsync($"metric:errors:{bucket}");

            batch.Execute();

            await Task.WhenAll(t3 is null
                ? new[] { t1, (Task)t2 }
                : new[] { t1, (Task)t2, t3 });

            // Thành công: reset circuit breaker
            _circuitBreaker.RecordSuccess();
        }
        catch (Exception ex)
        {
            // Metrics must never crash the request pipeline
            _logger.LogWarning(ex, "Failed to push metrics to Redis");

            // Ghi nhận thất bại để circuit breaker tính toán
            _circuitBreaker.RecordFailure();
        }
    }

    // Lớp circuit breaker nội bộ
    private class CircuitBreaker
    {
        private readonly int _failureThreshold;
        private readonly TimeSpan _cooldown;
        private readonly ILogger _logger;

        private int _failureCount;
        private DateTime _lastFailureTime;
        private readonly object _lock = new object();
        private bool _isOpen;

        public CircuitBreaker(int failureThreshold, TimeSpan cooldown, ILogger logger)
        {
            _failureThreshold = failureThreshold;
            _cooldown = cooldown;
            _logger = logger;
        }

        public bool AllowRequest()
        {
            lock (_lock)
            {
                if (!_isOpen)
                    return true;

                // Sau thời gian cooldown, chuyển sang trạng thái nửa mở (half-open)
                if (DateTime.UtcNow - _lastFailureTime > _cooldown)
                {
                    _isOpen = false;
                    _failureCount = 0;
                    _logger.LogInformation("Metrics circuit closed (recovered).");
                    return true;
                }

                return false;
            }
        }

        public void RecordSuccess()
        {
            lock (_lock)
            {
                _failureCount = 0;
                if (_isOpen)
                {
                    _isOpen = false;
                    _logger.LogInformation("Metrics circuit closed after success.");
                }
            }
        }

        public void RecordFailure()
        {
            lock (_lock)
            {
                if (_isOpen)
                    return;

                _failureCount++;
                _lastFailureTime = DateTime.UtcNow;

                if (_failureCount >= _failureThreshold)
                {
                    _isOpen = true;
                    _logger.LogWarning("Metrics circuit opened after {FailureCount} consecutive failures.", _failureCount);
                }
            }
        }
    }
}

File: slp/backend-nestjs/src/modules/metrics/redis-metrics-collector.ts
import { Injectable, Logger } from '@nestjs/common';
import { Redis } from 'ioredis';
import { ConfigService } from '@nestjs/config';
import { IMetricsCollector } from './metrics-collector.interface';

@Injectable()
export class RedisMetricsCollector implements IMetricsCollector {
  private readonly redis: Redis;
  private readonly logger = new Logger(RedisMetricsCollector.name);
  private failureCount = 0;
  private lastFailureTime = 0;
  private isOpen = false;
  private readonly failureThreshold = 5;
  private readonly cooldownMs = 60 * 1000;

  constructor(private configService: ConfigService) {
    const redisUrl = this.configService.get<string>('redis.url');
    if (!redisUrl) throw new Error("No redisUrl!");
    this.redis = new Redis(redisUrl);
  }

  private allowRequest(): boolean {
    if (!this.isOpen) return true;
    if (Date.now() - this.lastFailureTime > this.cooldownMs) {
      this.isOpen = false;
      this.failureCount = 0;
      this.logger.log('Metrics circuit closed (recovered).');
      return true;
    }
    return false;
  }

  private recordSuccess(): void {
    this.failureCount = 0;
    if (this.isOpen) {
      this.isOpen = false;
      this.logger.log('Metrics circuit closed after success.');
    }
  }

  private recordFailure(): void {
    if (this.isOpen) return;
    this.failureCount++;
    this.lastFailureTime = Date.now();
    if (this.failureCount >= this.failureThreshold) {
      this.isOpen = true;
      this.logger.warn(`Metrics circuit opened after ${this.failureCount} consecutive failures.`);
    }
  }

  async recordRequest(path: string, method: string, statusCode: number, latencyMs: number): Promise<void> {
    if (!this.allowRequest()) {
      this.logger.debug('Metrics circuit is open, skipping record.');
      return;
    }

    try {
      const bucket = new Date().toISOString().slice(0, 16); // yyyy-MM-ddTHH:mm
      const requestKey = `metric:requests:${bucket}`;
      const latencyKey = `metric:latency:${bucket}`;

      const pipeline = this.redis.pipeline();
      pipeline.incr(requestKey);
      pipeline.rpush(latencyKey, latencyMs.toFixed(2));
      if (statusCode >= 400) {
        pipeline.incr(`metric:errors:${bucket}`);
      }
      await pipeline.exec();

      this.recordSuccess();
    } catch (err) {
      this.logger.warn(`Failed to push metrics to Redis: ${err.message}`);
      this.recordFailure();
    }
  }
}